tools/autograd/templates/Functions.cpp

```diff
@@ -1800,44 +1800,151 @@ Tensor qr_backward(const std::vector<torch::autograd::Variable> &grads, const Te
+    }
+
+    Tensor grad_det = at::empty_like(self);
+    if (nonzero_det_indices[0].size(0) > 0) {  // invertible case
+      auto indexed_grad = grad.index(nonzero_det_indices);
+      auto indexed_det = det.index(nonzero_det_indices);
+      auto indexed_self = self.index(nonzero_det_indices);
+      grad_det.index_put_(/*indices=*/nonzero_det_indices,
+                          /*value=*/nonsingular_case_backward(indexed_grad, indexed_self, indexed_det));
+    }
+    if (zero_det_indices[0].size(0) > 0) {  // non-invertible case, uses SVD
+      auto indexed_grad = grad.index(zero_det_indices);
+      auto indexed_det = det.index(zero_det_indices);
+      auto indexed_self = self.index(zero_det_indices);
+      grad_det.index_put_(/*indices=*/zero_det_indices,
+                          /*value=*/singular_case_backward(indexed_grad, indexed_self, indexed_det));
+    }
+    return grad_det;
   }
 }
 
 Tensor logdet_backward(const Tensor & grad, const Tensor& self, const Tensor& logdet) {
-  auto logdet_val = logdet.item<double>();
-  if (logdet_val != -INFINITY /* det != 0, invertible */) {
-    return grad * self.inverse().t();
-  } else /* otherwise det = \prod(sigma) = 0, use svd */ {
+  auto singular_case_backward = [&](const Tensor& grad, const Tensor& self) -> Tensor {
     Tensor u, sigma, v;
     std::tie(u, sigma, v) = self.svd();
-    // backward det = \sum log(sigma)
-    auto gsigma = grad.div(sigma);
+    // logdet = \sum log(sigma)
+    auto gsigma = grad.unsqueeze(-1).div(sigma);
     return svd_backward({{}, gsigma, {}}, self, true, true, u, sigma, v);
+  };
+
+  auto nonsingular_case_backward = [&](const Tensor& grad, const Tensor& self) -> Tensor {
+    return unsqueeze_multiple(grad, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+  };
+
+  if (self.dim() == 2) {
+    if ((logdet != -INFINITY).item<uint8_t>()) {
+      return nonsingular_case_backward(grad, self);
+    } else {
+      return singular_case_backward(grad, self);
+    }
+  } else {
+    auto finite_logdet_indices = at::where(logdet != -INFINITY);
+    auto neginf_logdet_indices = at::where(logdet == -INFINITY);
+
+    if (finite_logdet_indices[0].size(0) == logdet.numel()) {  // all log determinants are finite (non-singular)
+      return nonsingular_case_backward(grad, self);
+    }
+
+    if (neginf_logdet_indices[0].size(0) == logdet.numel()) {  // all log determinants are -inf (singular)
+      return singular_case_backward(grad, self);
+    }
+
+    Tensor grad_logdet = at::empty_like(self);
+    if (finite_logdet_indices[0].size(0) > 0) {  // invertible case
+      auto indexed_grad = grad.index(finite_logdet_indices);
+      auto indexed_self = self.index(finite_logdet_indices);
+      grad_logdet.index_put_(/*indices=*/finite_logdet_indices,
+                          /*value=*/nonsingular_case_backward(indexed_grad, indexed_self));
+    }
+    if (neginf_logdet_indices[0].size(0) > 0) {  // non-invertible case, uses SVD
```

Comment:
this `if` condition will always be true

Reply:
I can’t see why this would be true always, could you please explain?

Reply:
Because if `neginf_logdet_indices[0].size(0) == 0`, then we have all singular case, which is covered by the `finite_logdet_indices[0].size(0) == logdet.numel()` case above. :)

Reply:
Got it, dumb me